notes/patching-system.md

```diff
@@ -0,0 +1,100 @@
+# The Patching System
+
+This document explains how Shrink Ray efficiently applies multiple changes in parallel.
+
+## The Problem
+
+When reducing, we often want to try many potential changes:
+- Delete each of 1000 lines
+- Delete each token
+- Replace each integer with a smaller value
+
+Trying these one-at-a-time is slow. But we can often apply multiple changes simultaneously if they don't conflict.
+
+## Patches Abstract Interface
+
+The `Patches[Patch, Target]` class defines how patches work:
+
+```python
+class Patches[Patch, Target]:
+    @property
+    def empty(self) -> Patch: ...           # The identity patch
+    def combine(self, *patches) -> Patch: ...  # Merge patches
+    def apply(self, patch, target) -> Target: ...  # Apply patch to target
+    def size(self, patch) -> int: ...       # How much the patch reduces
+```
+
+### Cuts (Most Common)
+
+`Cuts` patches are lists of `(start, end)` intervals to delete:
+```python
+patch = [(0, 5), (10, 15)]  # Delete bytes 0-5 and 10-15
+```
+
+Combining cuts merges their intervals. The `size` is total bytes deleted.
+
+### Other Patch Types
+
+- **SetPatches[T]**: A frozenset of items to remove (for JSON keys, etc.)
+- **ByteReplacement**: Dict mapping byte positions to replacement values
+
+## PatchApplier
+
+The core engine that applies patches in parallel with intelligent merging.
+
+### Merge Master Pattern
+
+Workers call `try_apply_patch()` in parallel. When a patch succeeds its interestingness test, it goes into a merge queue. One worker becomes the "merge master" and processes the queue:
```

Comment:
Worth noting that it takes out a lock, and when there is an active merge master all other workers have to wait for the merge master to complete or reject their patch.